server/src/main/java/org/elasticsearch/index/snapshots/blobstore/SlicedInputStream.java

```diff
@@ -99,4 +111,49 @@ public final int available() throws IOException {
         return stream == null ? 0 : stream.available();
     }
 
+    @Override
+    public boolean markSupported() {
+        return true;
+    }
+
+    @Override
+    public void mark(int readLimit) {
+        // We ignore readLimit since openSlice() can re-open previous InputStreams, and we can skip as many bytes as we'd like.
+        // According to JDK documentation, marking a closed InputStream should have no effect.
+        if (markSupported() && closed == false && numSlices > 0) {
+            if (initialized) {
+                markedSlice = nextSlice - 1;
+                markedSliceOffset = currentSliceOffset;
+            } else {
+                markedSlice = 0;
+                markedSliceOffset = 0;
+            }
+        }
+    }
+
+    @Override
+    public void reset() throws IOException {
+        if (markSupported()) {
+            // JDK documentation does not clarify if a closed InputStream (that has been previously marked) can be reset. We assume the same
+            // behavior specified for mark(), i.e., that reset on a closed InputStream has no effect.
```

Comment:
I'd prefer to fail on a closed stream here. Even assert false too.

Reply:
I made it throw if closed. Did not add an assertion, so I can test the behavior in the unit test. Feel free to tell me if the assertion is important to add.